### ISSUE TITLE
[ERROR] Worker (pid:25134) was sent SIGKILL! Perhaps out of memory?

### ISSUE THREAD

**Issue by UTSAV-44** [turboderp/exllamav2#556]
I am using A10 gpu with 24 gb ram for inferencing LLama3 .I am gunicorn with workers count 2 but It is giving Perhaps out of memory?.

**Comment by waterangel91** [comment]
I think worker as 2 will require double the memory requirement for GPU. 13*2>24

**Comment by UTSAV-44** [comment]
> I think worker as 2 will require double the memory requirement for GPU. 13*2>24

I have observed that 13gb is used when worker is 2.

**Comment by turboderp** [comment]
There is a known issue with safetensors that only shows up on some systems. Windows especially suffers from it, but I've seen it reported on some Linux systems as well. I think it has to do with memory mapping not working properly when you have too many files open at once, or something like that.

There is an option to bypass safetensors when loading models, which can be enabled with either `-fst` on the command line, setting the `EXLLAMA_FASTTENSORS` env variable, or setting `config.fasttensors` in Python.

**Comment by UTSAV-44** [comment]
Does it depends on NVIDIA driver version and cuda version.At present Driver Version: 535.183.01 and CUDA Version: 12.2

**Comment by turboderp** [comment]
No, it's an issue with safetensors and/or possibly the OS kernel. Try using one of the options above to see if it helps.

**Comment by UTSAV-44** [comment]
I tried with setting config.fasttensors = True ,but it does not work out.I tried using this in g4dn.xlarge instance but the model is not loading .

**Comment by turboderp** [comment]
Can you share the code that fails? The config option has to be set after `config.prepare()` is called but before `model.load()`.

**Comment by UTSAV-44** [comment]
config = ExLlamaV2Config(model_dir)
		config.fasttensors = True
		self.model = ExLlamaV2(config)
		
		self.cache = ExLlamaV2Cache_Q4(self.model, max_seq_len=256*96, lazy=True)  
		self.model.load_autosplit(self.cache, progress=True)

		print("Loading tokenizer...")
		self.tokenizer = ExLlamaV2Tokenizer(config)

		self.generator = ExLlamaV2DynamicGenerator(
			model=self.model,
			cache=self.cache,
			tokenizer=self.tokenizer,
		)

		self.generator.warmup()
		
I am running it on kubernetes with g5.xlarge gpu instance.

**Comment by turboderp** [comment]
I'm not sure there's any way to prevent PyTorch from using a lot of virtual memory. But just out of interest, what do you get from the following?

```sh
cat /proc/sys/vm/overcommit_memory
ulimit -v
```

**Comment by UTSAV-44** [comment]
for the command 
cat /proc/sys/vm/overcommit_memory I got 1
ulimit -v I got unlimited

**Comment by turboderp** [comment]
I'm not sure about the implications actually, but I *think* you might want to try changing the overcommit mode.

`sudo sysctl vm.overcommit_memory=0`

or

`sudo sysctl vm.overcommit_memory=2`

:shrug:

**Comment by brthor** [comment]
@turboderp 
I am also encountering a sigkill on a system with 128gb system memory.

This seems to occur during the first generation run, but GPU utilization is low while the system memory use increases.

If OOM is not encountered, the system memory then resets to a more normal value, like 2-4gb, and the gpu utilization goes to 100% and generation proceeds as normal.

Reducing the cache size solves the issue but reduces generation speed.

The error occurs with 
`vm.overcommit_memory=0`
and
`vm.overcommit_memory=1`

cfg fasttensors does not help, but I only have .safetensors files.

This is running in a docker container with pytorch 2.3.1, installed from the `exllamav2-0.1.8+cu121.torch2.3.1-cp310-cp310-linux_x86_64.whl` wheel.
